src/abstractdataframe/abstractdataframe.jl

```diff
@@ -1450,6 +1450,150 @@ julia> unique!(df)  # modifies df
 """
 (unique, unique!)
 
+"""
+    completecombinations(df::AbstractDataFrame, indexcols; allcols::Bool=false,
+                         allowduplicates::Bool=false, fill=missing)
+
+    Generate all combinations of `levels` (including `missing` value if
+    present) of `indexcols` in `df` data frame in the order of `levels`.
+
+    If `allcols` is `false` (the default) in the resulting data frame only the
+    unique combinations of `indexcols` are included (without duplicates).
+
+    If `allcols` is `true` all columns from `df` are added to the resulting data
+    frame (the `indexcols` are the first columns stored in the result). For the
+    combinations of `indexcols` these columns take `fill` value (`missing`) by
+    default.
+
+    If `allowduplicates` is `false` (the default) `indexcols` may only contain
+    unique combinations of `indexcols` values. If `allowduplicates` is `true`
+    duplicates are allowed. They are not repeated if `allcols` is `false`
+    only unique combinations are produced then, but if `allcols` is `true`
+    the duplicates are included.
```

Comment:
This behavior differs from dplyr's `complete`, which retains duplicates. Any particular reason to drop them? Or should we allow three options about how duplicates should be handled?

Reply:
It is a consequence of the current design. I.e. it is natural to drop them in the way it is implemented.
But we can discuss what to do.

Now I am thinking that maybe we do not need the `allcols` kwarg and can simplify the API. Maybe we should have `allcols=true` always and expect from the user to pass a data frame only with columns that are wanted in the output?